controlplane/pkg/nsm/nsm_heal.go

```diff
@@ -153,27 +219,55 @@ func (srv *networkServiceManager) requestOrClose(logPrefix string, ctx context.C
 		logrus.Infof("%v Heal: Connection recovered: %v", logPrefix, connection)
 	}
 }
-func (srv *networkServiceManager) waitAnyNSE(clientConnection *model.ClientConnection) {
+
+
+func (srv *networkServiceManager) waitSpecificNSE(ctx context.Context, clientConnection *model.ClientConnection) bool {
+	discoveryClient, err := srv.serviceRegistry.DiscoveryClient()
+	if err != nil {
+		logrus.Errorf("Failed to connect to Registry... %v", err)
+		// Still try to recovery
+		return false
+	}
+
 	st := time.Now()
+
+	networkService := clientConnection.Endpoint.NetworkService.Name
+	nseRequest := &registry.FindNetworkServiceRequest{
+		NetworkServiceName: networkService,
+	}
+
+	defer func() {
+		logrus.Infof("Complete Waiting for Remote NSE/NSMD with network service %s. Since elapsed: %v", networkService, time.Since(st))
+	}()
+
 	for {
-		logrus.Infof("NSM: LocalNSE: Waiting for NSE with network service %s. Since elapsed: %v", clientConnection.Endpoint.NetworkService.Name, time.Since(st))
-		ignored := map[string]*registry.NSERegistration{}
-		nsmConnection := srv.newConnection(clientConnection.Request)
-		ep, err := srv.getEndpoint(context.Background(), nsmConnection, ignored)
-		if err == nil && ep != nil {
-			// We could call requires since we have some Endpoint to check with.
-			break
+		logrus.Infof("NSM: RemoteNSE: Waiting for NSE with network service %s NSE %v. Since elapsed: %v", networkService, clientConnection.Endpoint.NetworkserviceEndpoint.EndpointName, time.Since(st))
+
+		endpointResponse, err := discoveryClient.FindNetworkService(ctx, nseRequest)
```

Comment:
`endpointResponse` is a reference, it will panic if it'll be nil

Reply:
Did you use any kind of tool to find this places?